kvscheduler/api/kv_scheduler_api.go

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+	"github.com/ligato/cn-infra/idxmap"
+)
+
+// KeySelector is used to filter keys.
+type KeySelector func(key string) bool
+
+// ValueType is one of: Object, Action or Property.
+type ValueType int
+
+const (
+	// Object is something that exists.
+	//
+	// Objects are typically base (non-derived) values, but it is allowed
+	// to derive (sub)object from another object. It is actually allowed
+	// to derive/ any value type from an object.
+	// Without associated descriptor, object cannot be created (build + add)
+	// via NB transaction (considered unimplemented) - it can only be pushed
+	// from SB.
+	// Descriptor is allowed to associate metadata with created non-derived
+	// object values and let the scheduler to maintain and expose the mapping
+	// between value labels and metadata (possibly extended with secondary
+	// indices).
+	Object ValueType = iota
+
+	// Action is something to be executed when dependencies are met and reverted
+	// when it is no longer the case. Actions are often derived from objects.
+	//
+	// Like objects, actions can be base or derived values, but it is not allowed
+	// to further derive more values from actions. Also, actions are not expected
+	// to have metadata associated with them.
+	// Without associated descriptor, there are obviously no actions to execute,
+	// thus un-described action values are never create - they can only be pushed
+	// (as already executed) from SB.
+	Action
+
+	// Property is something an object has and other kv pairs may depend on.
+	//
+	// Property is always derived from object - it cannot be created/pushed as
+	// a base value. Property cannot have descriptor associated with it
+	// (i.e. no methods to execute, no dependencies or further derived values,
+	// no associated metadata).
+	Property
+)
+
+// String converts ValueType to string.
+func (vt ValueType) String() string {
+	switch vt {
+	case Object:
+		return "object"
+	case Action:
+		return "action"
+	default:
```

Comment:
This case should match if `vt==Property` and use default for all other values. Otherwise something like this could occur:

```go
v := ValueType(-1)
v.String() // will print "property"
if v == Property {
 // will not be executed
}
```

This might be harder to debug.